Add APIs and initial plugins for GUI support and fit Windows major APIs to current design flow

atcuno opened the pull request:
This PR accomplishes the final major API additions and overhauls needed before we can get to the parity release. This was a collaboration with myself and @dgmcdona. With these changes, only the new GUI-related windows, messagehooks, and rawinputdevicemonitors plugins are remaining before parity. These will be short and build off the APIs added in this PR.

From a high level, this PR covers:

1)	Changing the major Windows APIs (listing processes, modules, and registry hives) to  fit our current design flow (sending in the kernel module name instead of different pieces), which also matches Linux and fixes a painful amount of bugs that we found while getting the GUI API developed.

2)	Adding the GUI API, which required adding the ability to pool scan across different layers and types, as well as the GUI-specific APIs and building blocks. This also led to fixing many bugs along the way while getting it all working.

3)	Adding the initial GUI plugin (windowstations) and then the next logical GUI step (desktops of Windows) through both re-using the workstation scanning (desktops plugin) as well as direct desktop scanning (deskscan). Deskscan shows how the remaining GUI plugins will re-use the GUI-specific scanning infrastructure. Desktops shows how the new APIs can be integrated easily into future GUI work outside the core team.

All these changes were tested through multiple runs against the full Windows sample set (411 samples covering XP-11 and 32/64 bit).

The deeper details:

The three core Windows APIs (list_processes, list_modules, hivelist) all used the flawed/old way of sending parameters to reference the kernel symbol table and layer. This led to all the follow on plugins also following this pattern, which led to a significant amount of bugs as plugins and APIs would 1) confuse layers and tables between the kernel and processes, and 2) these issues was *extremely* hard to debug as the code used crappy names like `layer_name`, which is ambiguous in code paths dealing with multiple layers, and names like `symbol_table` which were actually the name of the table and not the table itself. All `symbol_table` named variables in the affected code paths were changed to `symbol_table_name`. With these API changes, code must be 100% explicit when it wants to reference the kernel, and being able to reference kernel.layer_name is much harder to break then accidently sending in the wrong (badly named) variable.
To remedy this overall situation with the least amount of changes for one PR, David (@dgmcdona) and I changed the three core APIs to take in only the kernel module name, which then led to the downstream using APIs doing the same. We only touched the APIs directly needed to make the GUI code work, not other APIs buried in different plugins. Given that this encompassed the core 3 APIs (processes, modules, hives), we are quite certain we will prevent these bad patterns from showing up in future plugins. These changes makes, like Linux, to where only functions that need the kernel object construct it, instead of it being passed around everywhere. This immediately called out several bugs in existing APIs. 

The effects of these changes were pretty minimal in most plugins as they would call list_processes or list_modules from `run` or `_generator`. Only a few plugins needed major changes to accommodate this, which we fixed.

We then went through and changed all needed requirement dependencies versions as well as bumped versions in plugins that changed their API signatures. With these changes in place, all the new GUI APIs follow the convention correctly from the start to avoid future overhauls.

As mentioned before, the branch in its current state (and a few previous ones) were tested against all 411 Windows samples in our internal set. In the final run, the only backtraces triggered were unrelated to our changes, and I made separate tickets for them with the parity label added.

Fixes #1504 #1636

dgmcdona committed:
Windows: Adds windows GUI vtypes

dgmcdona committed:
Windows: Adds windows OS version checks

dgmcdona committed:
Windows: Adds GUI plugin

atcuno committed:
Windows DeskScan: Update pool extension

This updates the Windows pool extension with a new get_name() method,
and changes the method signature for another. We'll need to figure out
if there is a good way to version extension classes.

Deskscan requires this in order to use get_name with an alternate
(non-kernel) symbol table.

dgmcdona committed:
Windows WindowStations: Fix issue with native types

atcuno committed:
Windows Scheduled Tasks: Prevent backtraces

This fixes an `AttributeError` that can crop up when the actionset's
context is `None`.

dgmcdona committed:
Windows Versions: Sort version checks for readability

dgmcdona committed:
Windows Versions: Add version check for win10 17735 or later

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -252,17 +254,14 @@ def get_amcache_hive(
-                symbol_table=kernel.symbol_table_name,
+                context,
+                interfaces.configuration.path_join(config_path, "hivelist"),
+                kernel_module_name,
```
I'd prefer that we pass these as keyword arguments, rather than just positional, to avoid possible confusion and so on.  The context is almost always first but the other ideally would be keyword parameters.

ikelos replied:
```suggestion
                context=context,
                base_config_path=interfaces.configuration.path_join(config_path, "hivelist"),
                kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/layers/registry.py:
```diff
@@ -66,7 +66,7 @@ def __init__(
         # Win10 17063 introduced the Registry process to map most hives.  Check
```
Any chance of removing the use of the plugin from the layer?  This had a tendency to cause import loops and is generally bad form, plus I don't know that all the functionality from list_processes is needed?  I also don't know where the dependencies for this layer are specified, since I can't see them on it anywhere?

ikelos commented on volatility3/framework/layers/registry.py:
```diff
@@ -66,7 +66,7 @@ def __init__(
         # Win10 17063 introduced the Registry process to map most hives.  Check
         # if it exists and update RegistryHive._base_layer
         for proc in pslist.PsList.list_processes(
-            self.context, self.config["base_layer"], self.config["nt_symbols"]
```
So, the registry layer has a `kernel_module_name` option that it must be provided with when instantiating it?  I'd expect it to appear in a `requirements` list or something?  I'm sure it works, I'd just want to make sure the how is clearly documented.

atcuno replied:
This was existing before, all we did was change it to match the new `list_processes` parameter list. If you want the registry layer not walking the list, we can examine that but can it be in a seperate PR? I can make a new ticket with the parity label for Dave to look into. I also know that on Win10+ everything registry hive related is in its own Registry process, which has its own layer so that would be my first guess why its calling there.

ikelos replied:
Errr, so walking the list is ok, pulling in pslist to do it is the problem.  I *believe* it should just be enumerating a list (looked like 3-4 lines when I glanced at it last) which I'd prefer to have duplicated than pulling in the whole plugin just to do it.  I also don't know where the config options came from (because they shouldn't be part of the registry layer's requirements, or there should be some explicit code listing them somewhere (like a `get_requirements` or something).  Just seems a bit off, so yeah, if you could raise a ticket I'd appreciate it.  Doesn't impact this PR particularly, just something we need to get sorted out.

ikelos commented on volatility3/framework/plugins/windows/consoles.py:
```diff
@@ -823,10 +824,9 @@ def get_console_settings_from_registry(
-            base_config_path=config_path,
-            layer_name=kernel_layer_name,
-            symbol_table=kernel_symbol_table_name,
+            context,
```
Again, keyword args rather than positional please

atcuno replied:
I personally find this leads to worse code, like people using crappy variable names and hoping the parameter name clarifies. We can change them back though if its a deal breaker though?

ikelos replied:
It sounds like there's pros and cons.  Bad variable names we can at least see and vet at the PR stage.  Accidentally swapping two variable names because you thought you remembered the precise ordering feels more difficult to spot in a PR (particularly if they're the same types).  It does have the downside of making it more difficult to rename the variables later (because every caller essentially has a copy of the function signature in it), but that feels like what should be an uncommon minor case.  I think I would like them in (where the parameters are more than one line, if the linter doesn't multi-line them without the variable names, then it's fine).

atcuno replied:
Yea this works. David is overhauling these places now and will update the PR with the names back.

ikelos commented on volatility3/framework/plugins/windows/consoles.py:
```diff
@@ -946,9 +943,8 @@ def run(self):
-                    context=self.context,
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
```
Not sure why you took out existing keyword args, they can only help with clarity?

ikelos commented on volatility3/framework/plugins/windows/debugregisters.py:
```diff
@@ -35,13 +35,13 @@ def get_requirements(cls) -> List:
             ),
-            requirements.VersionRequirement(
-                name="threads", component=threads.Threads, version=(1, 0, 0)
+            requirements.PluginRequirement(
```
Plugin and Version requirement are identical but with plugin rather than component?  I think the idea was to make PluginRequirements the more generic VersionRequirement, and slowly shift all of them over to that.  This puts that in reverse, could we try to eventually phase out PluginRequirements please?

atcuno replied:
Ok that works. I will switch any of the plugin ones over.

ikelos replied:
You can just do the ones you're adding (or tried to change).  Any that were already there we'll pick up in a single big sweep of the whole codebase (we'll need an issue to remember to do it though).  5:)

atcuno replied:
This general issue now here: https://github.com/volatilityfoundation/volatility3/issues/1644

but we will resolve it in this PR also

ikelos commented on volatility3/framework/plugins/windows/cachedump.py:
```diff
@@ -169,13 +169,11 @@ def run(self):
             self.config_path,
-            kernel.layer_name,
-            kernel.symbol_table_name,
+            self.config["kernel"],
```
keyword arg

ikelos replied:
```suggestion
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/deskscan.py:
```diff
@@ -0,0 +1,81 @@
+    """Scans for the Desktop instances of each Window Station"""
+
+    _required_framework_version = (2, 0, 0)
+
```
This has class methods, but no `_version` defined?  It should be `(1, 0, 0)` please

atcuno replied:
Good spot.

ikelos commented on volatility3/framework/plugins/windows/deskscan.py:
```diff
@@ -0,0 +1,81 @@
+                description="Windows kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+            requirements.PluginRequirement(
```
Let's try to go for the more generic VersionRequirements I think?

atcuno replied:
Resolving as now a part of: https://github.com/volatilityfoundation/volatility3/issues/1644

ikelos commented on volatility3/framework/plugins/windows/desktops.py:
```diff
@@ -0,0 +1,91 @@
+                description="Windows kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+            requirements.PluginRequirement(
```
VersionRequirement please

ikelos replied:
It might be better to just do a full sweep at some point.  I was expecting them all to shift over, over time, but it seems there's confusion.  Happy to discuss it if people think having the second (aliased) name for VersionRequirement is useful, but for me they're both accurate and it just muddies the water.  We should probably add a deprecation warning when using them...

ikelos replied:
I probably won't bother pointing them out again.

atcuno replied:
Resolving as now a part of https://github.com/volatilityfoundation/volatility3/issues/1644

ikelos commented on volatility3/cli/volshell/windows.py:
```diff
@@ -38,11 +38,7 @@ def change_process(self, pid=None):
-                self.context, self.current_layer, self.current_symbol_table
-            )
-        )
+        return list(pslist.PsList.list_processes(self.context, self.config["kernel"]))
```
I don't think `self.config` is the right value to use for this?  I think it's supposed to be `self.current_kernel_name` so people have the opportunity to change it.

atcuno replied:
Ok I will fix this too.

ikelos commented on volatility3/framework/plugins/windows/debugregisters.py:
```diff
@@ -117,11 +117,7 @@ def _generator(
+        procs = pslist.PsList.list_processes(self.context, self.config["kernel"])
 
         for proc in procs:
             for thread in threads.Threads.list_threads(kernel, proc):
```
This looks like an API where we hand in a kernel object rather than a context and kernel_name.  Possibly for one a future PR?

atcuno replied:
Agreed it is on the list for after this PR goes it. For this one, we were trying to keep it to just the APIs that the GUI code went through (listing process, modules, hives)

atcuno replied:
Resolving as now tracked here: https://github.com/volatilityfoundation/volatility3/issues/1648

ikelos commented on volatility3/framework/plugins/windows/handles.py:
```diff
@@ -393,9 +393,8 @@ def run(self):
-                layer_name=kernel.layer_name,
-                symbol_table=kernel.symbol_table_name,
+                self.context,
+                self.config["kernel"],
```
```suggestion
                context=self.context,
                kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/getsids.py:
```diff
@@ -217,15 +215,13 @@ def _generator(self, procs):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],

```

ikelos commented on volatility3/framework/plugins/windows/getsids.py:
```diff
@@ -101,14 +101,12 @@ def lookup_user_sids(self) -> Dict[str, str]:
-            symbol_table=kernel.symbol_table_name,
+            self.context,
+            self.config_path,
+            self.config["kernel"],
```
```suggestion
            context=self.context,
            base_config_path=self.config_path,
            kernel_module_name=self.config["kernel"],

```

ikelos commented on volatility3/framework/plugins/windows/getservicesids.py:
```diff
@@ -69,18 +69,16 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
-            symbol_table=kernel.symbol_table_name,
+            self.context,
+            self.config_path,
+            self.config["kernel"],
```
```suggestion
            context=self.context,
            base_config_path=self.config_path,
            kernel_module_name=self.config["kernel"],

```

ikelos commented on volatility3/framework/plugins/windows/envars.py:
```diff
@@ -228,9 +225,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],

```

ikelos commented on volatility3/framework/plugins/windows/envars.py:
```diff
@@ -58,13 +58,11 @@ def _get_silent_vars(self) -> List[str]:
-            symbol_table=kernel.symbol_table_name,
+            self.context,
+            self.config_path,
+            self.config["kernel"],
```
```suggestion
            context=self.context,
            base_config_path=self.config_path,
            kernel_module_name=self.config["kernel"],

```

ikelos commented on volatility3/framework/plugins/windows/dumpfiles.py:
```diff
@@ -373,8 +372,7 @@ def run(self):
                 self.context,
-                kernel.layer_name,
-                kernel.symbol_table_name,
+                self.config["kernel"],
```
```suggestion
                kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/dlllist.py:
```diff
@@ -222,9 +217,8 @@ def run(self):
-                layer_name=kernel.layer_name,
-                symbol_table=kernel.symbol_table_name,
+                self.context,
+                self.config["kernel"],
```
```suggestion
                context=self.context,
                kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/direct_system_calls.py:
```diff
@@ -350,12 +351,15 @@ def get_tasks_to_scan(
-            layer_name=layer_name,
-            symbol_table=symbol_table_name,
+            context,
+            kernel_module_name,
```
```suggestion
            context=context,
            kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/plugins/windows/consoles.py:
```diff
@@ -946,9 +943,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/consoles.py:
```diff
@@ -823,10 +824,9 @@ def get_console_settings_from_registry(
-            symbol_table=kernel_symbol_table_name,
+            context,
+            config_path,
+            kernel_module_name,
```
```suggestion
            context=context,
            base_config_path=config_path,
            kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/plugins/windows/cmdscan.py:
```diff
@@ -373,9 +370,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/cmdline.py:
```diff
@@ -99,16 +100,14 @@ def _generator(self, procs):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/memmap.py:
```diff
@@ -109,9 +108,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/malfind.py:
```diff
@@ -257,9 +256,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/lsadump.py:
```diff
@@ -209,13 +209,11 @@ def _generator(
             self.config_path,
-            kernel.layer_name,
-            kernel.symbol_table_name,
+            self.config["kernel"],
```
```suggestion
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/ldrmodules.py:
```diff
@@ -121,9 +120,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/joblinks.py:
```diff
@@ -36,16 +36,18 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
         for proc in pslist.PsList.list_processes(
-            self.context, kernel.layer_name, kernel.symbol_table_name
+            self.context,
+            self.config["kernel"],
```
```suggestion
            context=self.context,
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/iat.py:
```diff
@@ -139,9 +137,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/hollowprocesses.py:
```diff
@@ -215,9 +214,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/hashdump.py:
```diff
@@ -593,12 +593,10 @@ def run(self):
             self.config_path,
-            kernel.layer_name,
-            kernel.symbol_table_name,
+            self.config["kernel"],
```
```suggestion
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/modscan.py:
```diff
@@ -53,34 +55,33 @@ def get_requirements(cls):
+            kernel.symbol_table_name, [b"MmLd"]
         )
 
         for result in poolscanner.PoolScanner.generate_pool_scan(
```
Are we keeping `generate_pool_scan` intentionally with layer_name and symbol_table_name, or was this just to keep the PR small?

atcuno replied:
For this PR, yes, we kept `generate_pool_scan` to match its existing interface then made extended version call into it. In a new PR/effort after this one we will change it to be just the kernel module name for the kernel-only scanners (everything but the GUI code so far), and then update those scanners. That will be a very clean PR overall.

atcuno replied:
Resolving as now part of here: https://github.com/volatilityfoundation/volatility3/issues/1646

ikelos commented on volatility3/framework/plugins/windows/modules.py:
```diff
@@ -92,13 +92,12 @@ def _generator(self):
                 self.get_session_layers(
-                    self.context, kernel.layer_name, kernel.symbol_table_name
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/debugregisters.py:
```diff
@@ -35,13 +35,13 @@ def get_requirements(cls) -> List:
-            requirements.VersionRequirement(
-                name="threads", component=threads.Threads, version=(1, 0, 0)
+            requirements.PluginRequirement(
+                name="threads", plugin=threads.Threads, version=(2, 0, 0)
```
```suggestion
            requirements.VersionRequirement(
                name="threads", component=threads.Threads, version=(2, 0, 0)
```

ikelos commented on volatility3/framework/plugins/windows/modules.py:
```diff
@@ -176,20 +174,20 @@ def get_session_layers(
-            layer_name=layer_name,
-            symbol_table=symbol_table,
+            context,
+            kernel_module_name,
```
```suggestion
            context=context,
            kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/plugins/windows/modules.py:
```diff
@@ -235,8 +235,53 @@ def get_session_layers(
+        ):
+            sessions[session_id] = proc_layer_name
+
+        return sessions
```
```suggestion
        return dict(cls._do_get_session_layers(
            context, kernel_module_name, pids
        ))
```

ikelos commented on volatility3/framework/plugins/windows/modules.py:
```diff
@@ -268,26 +313,29 @@ def find_session_layer(
+        kernel = context.modules[kernel_module_name]
+
+        kvo = context.layers[kernel.layer_name].config.get(
+            "kernel_virtual_offset", None
+        )
         if not kvo:
             raise ValueError(
                 "Intel layer does not have an associated kernel virtual offset, failing"
             )
-        ntkrnlmp = context.module(symbol_table, layer_name=layer_name, offset=kvo)
+        ntkrnlmp = context.module(
+            kernel.symbol_table_name, layer_name=kernel.layer_name, offset=kvo
+        )
```
Errr, isn't this all just ntkrnlmp = kernel?

atcuno replied:
I just modified existing code. We can update to match to use the `ntkrnlmp` though

ikelos replied:
Well, it just looked like it was using the kernel data to figure out what kernel was!  Seemed unnecessary when you can just use kernel.  5;P

ikelos replied:
Still outstanding, the lines above are to derive `ntkrnlmp` from `kernel` when they're actually just the same thing.

ikelos commented on volatility3/framework/plugins/windows/orphan_kernel_threads.py:
```diff
@@ -57,19 +59,16 @@ def list_orphan_kernel_threads(
         collection = ssdt.SSDT.build_module_collection(
-            context, layer_name, symbol_table_name
+            context,
+            kernel_module_name,
```
```suggestion
            context=context,
            kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/plugins/windows/pedump.py:
```diff
@@ -224,7 +237,11 @@ def _generator(self):
+                self.config["kernel"],
+                pe_table_name,
+                self.open,
+                self.config["base"],
```
```suggestion
                context=self.context,
                kernel_module_name=self.config["kernel"],
                pe_table_name=pe_table_name,
                open_method=self.open,
                base=self.config["base"],
```

ikelos commented on volatility3/framework/plugins/windows/poolscanner.py:
```diff
@@ -331,11 +362,12 @@ def builtin_constraints(
         context: interfaces.context.ContextInterface,
-        layer_name: str,
-        symbol_table: str,
+        kernel_layer_name: str,
```
Separate kernel values, rather than just `kernel_module_name`?

ikelos replied:
Given this is a new call, we should be setting it up the way we want it to be called?

atcuno replied:
See above, we will do that in a new PR after this one merged that overhauls the poolscanners or this PR would have gone out of hand.

atcuno replied:
Resolving as covered here now: https://github.com/volatilityfoundation/volatility3/issues/1646

ikelos commented on volatility3/framework/plugins/windows/privileges.py:
```diff
@@ -120,9 +119,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],

```

ikelos commented on volatility3/framework/plugins/windows/processghosting.py:
```diff
@@ -95,9 +94,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/pslist.py:
```diff
@@ -206,32 +208,37 @@ def filter_func(x):
-        ntkrnlmp = context.module(symbol_table, layer_name=layer_name, offset=kvo)
+        ntkrnlmp = context.module(
+            kernel.symbol_table_name, layer_name=kernel.layer_name, offset=kvo
+        )
```
Again, all this could be in comments, or removed, since it's essentially `ntkrnlmp = kernel`

ikelos commented on volatility3/framework/plugins/windows/pslist.py:
```diff
@@ -273,8 +280,7 @@ def _generator(self):
             self.context,
-            kernel.layer_name,
-            kernel.symbol_table_name,
+            self.config["kernel"],
```
```suggestion
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/registry/hivelist.py:
```diff
@@ -95,8 +97,7 @@ def _generator(self) -> Iterator[Tuple[int, Tuple[int, str]]]:
                         self.config_path,
-                        layer_name=kernel.layer_name,
-                        symbol_table=kernel.symbol_table_name,
+                        self.config["kernel"],
```
```suggestion
                        kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/registry/hivelist.py:
```diff
@@ -148,20 +148,24 @@ def list_hives(
+                        context,
+                        kernel.layer_name,
+                        kernel.symbol_table_name,
+                        filter_string,
```
```suggestion
                        context=context,
                        layer_name=kernel.layer_name,
                        symbol_table=kernel.symbol_table_name,
                        filter_string=filter_string,
```

ikelos commented on volatility3/framework/plugins/windows/registry/printkey.py:
```diff
@@ -240,17 +240,14 @@ def _printkey_iterator(
             self.config_path,
-            layer_name=layer_name,
-            symbol_table=symbol_table,
+            self.config["kernel"],
```
```suggestion
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/registry/userassist.py:
```diff
@@ -295,18 +295,16 @@ def _generator(self):
-            symbol_table=kernel.symbol_table_name,
+            self.context,
+            self.config_path,
+            self.config["kernel"],
```
```suggestion
            context=self.context,
            base_config_path=self.config_path,
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -1135,17 +1135,14 @@ def get_software_hive(
-                symbol_table=kernel.symbol_table_name,
+                context,
+                interfaces.configuration.path_join(config_path, "hivelist"),
+                kernel_module_name,
```
```suggestion
                context=context,
                base_config_path=interfaces.configuration.path_join(config_path, "hivelist"),
                kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/plugins/windows/sessions.py:
```diff
@@ -39,16 +39,14 @@ def get_requirements(cls):
             self.context,
-            kernel.layer_name,
-            kernel.symbol_table_name,
+            self.config["kernel"],
```
```suggestion
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/skeleton_key_check.py:
```diff
@@ -672,9 +670,8 @@ def run(self):
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/ssdt.py:
```diff
@@ -64,17 +66,13 @@ def build_module_collection(
             module_name = os.path.splitext(module_name_with_ext)[0]
 
-            symbol_table_name = None
-            if module_name in constants.windows.KERNEL_MODULE_NAMES:
```
This code would have... failed to set a symbol table name,if the module_name wasn't in the kernel.  Is there ever a time where it would want a different symbol table name?  This is likely just passed to the module, but I can imagine new created modules should carry their own symbol table name, rather than the kernel's?

atcuno replied:
Yes, the code was broken before (something we discovered while working on this one). Kernel modules only come from the kernel so its okay the new way. Remember this API is just to get the initial kernel module list, which those symbols and types are in the kernel themselves. If a kernel module wants to use its own then it gets made seperately. This `build_module_collection` is related to the APIs for mapping addresses (like function pointers) to their hosting module.

ikelos commented on volatility3/framework/plugins/windows/ssdt.py:
```diff
@@ -84,9 +82,9 @@ def build_module_collection(
         collection = self.build_module_collection(
-            self.context, layer_name, kernel.symbol_table_name
+            self.context,
+            self.config["kernel"],
```
```suggestion
            context=self.context,
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/strings.py:
```diff
@@ -68,12 +71,10 @@ def _generator(self) -> Generator[Tuple, None, None]:
             self.context,
-            kernel.layer_name,
-            kernel.symbol_table_name,
+            self.config["kernel"],
```
```suggestion
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/suspicious_threads.py:
```diff
@@ -35,7 +35,13 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
+                name="pslist", plugin=pslist.PsList, version=(3, 0, 0)
+            ),
+            requirements.PluginRequirement(
+                name="threads", plugin=threads.Threads, version=(2, 0, 0)
```
```suggestion
            requirements.VersionRequirement(
                name="pslist", component=pslist.PsList, version=(3, 0, 0)
            ),
            requirements.VersionRequirement(
                name="threads", component=threads.Threads, version=(2, 0, 0)
```

ikelos commented on volatility3/framework/plugins/windows/suspicious_threads.py:
```diff
@@ -132,9 +138,8 @@ def _enumerate_processes(
-            layer_name=kernel.layer_name,
-            symbol_table=kernel.symbol_table_name,
+            self.context,
+            self.config["kernel"],
```
```suggestion
            context=self.context,
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/svclist.py:
```diff
@@ -60,37 +62,37 @@ def _get_exe_range(cls, proc) -> Optional[Tuple[int, int]]:
-            layer_name=layer_name,
-            symbol_table=symbol_table,
+            context,
+            kernel_module_name,
```
```suggestion
            context=context,
            kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/plugins/windows/svcscan.py:
```diff
@@ -144,15 +144,13 @@ def _create_service_table(
-            symbol_table=symbol_table,
+            context,
+            interfaces.configuration.path_join(config_path, "hivelist"),
+            kernel_module_name,
```
```suggestion
            context=context,
            base_config_path=interfaces.configuration.path_join(config_path, "hivelist"),
            kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/plugins/windows/svcscan.py:
```diff
@@ -300,9 +299,8 @@ def service_scan(
-            layer_name=layer_name,
-            symbol_table=symbol_table,
+            context,
+            kernel_module_name,
```
```suggestion
            context=context,
            kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/plugins/windows/threads.py:
```diff
@@ -59,16 +59,18 @@ def list_threads(
-            layer_name=layer_name,
-            symbol_table=symbol_table_name,
+            context,
+            kernel_module_name,
```
```suggestion
            context=context,
            kernel_module_name=kernel_module_name,
```

ikelos commented on volatility3/framework/plugins/windows/timers.py:
```diff
@@ -122,15 +122,18 @@ def list_timers(
         collection = ssdt.SSDT.build_module_collection(
-            self.context, kernel.layer_name, kernel.symbol_table_name
+            self.context,
+            self.config["kernel"],
```
```suggestion
            context=self.context,
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/timers.py:
```diff
@@ -122,15 +122,18 @@ def list_timers(
+        # FIXME - the list_timers API is gross. Fix after GUI merge
         for timer in self.list_timers(
-            self.context, self.config["kernel"], layer_name, symbol_table
+            self.context,
+            self.config["kernel"],
+            kernel.layer_name,
+            kernel.symbol_table_name,
```
```suggestion
            context=self.context,
            kernel_module_name=self.config["kernel"],
            layer_name=kernel.layer_name,
            symbol_table_name=kernel.symbol_table_name,
```

ikelos replied:
Actually, this should be cleaned up at some point.  There seems to be a chain of plugins (`KPCRs.list_kpcrs`, and `Timers.list_timers` that take in a kernel module *and* a layer and symbol_table.  Unless there's value in allowing such flexibility, they should probably be condensed.  Maybe not in this PR, but then a new issue should be filed please.

atcuno replied:
Yes, its on our list for right after this PR. I left a comment about it in timers.

atcuno replied:
This is now tracked here: https://github.com/volatilityfoundation/volatility3/issues/1642

ikelos commented on volatility3/framework/plugins/windows/unhooked_system_calls.py:
```diff
@@ -153,18 +149,15 @@ def _gather_code_bytes(
             self.config_path,
-            kernel.layer_name,
-            kernel.symbol_table_name,
+            self.config["kernel"],
```
```suggestion
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/vadinfo.py:
```diff
@@ -294,9 +292,8 @@ def run(self) -> renderers.TreeGrid:
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/vadregexscan.py:
```diff
@@ -111,11 +111,9 @@ def _generator(self, regex_pattern, procs):
             self.context,
-            kernel.layer_name,
-            kernel.symbol_table_name,
+            self.config["kernel"],
```
```suggestion
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/vadwalk.py:
```diff
@@ -84,9 +83,8 @@ def run(self) -> renderers.TreeGrid:
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
+                    self.config["kernel"],
```
```suggestion
                    context=self.context,
                    kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/vadyarascan.py:
```diff
@@ -53,18 +53,15 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
-            layer_name=kernel.layer_name,
-            symbol_table=kernel.symbol_table_name,
+            self.context,
+            self.config["kernel"],
```
```suggestion
            context=self.context,
            kernel_module_name=self.config["kernel"],
```

ikelos commented on volatility3/framework/plugins/windows/windowstations.py:
```diff
@@ -0,0 +1,236 @@
+            context,
+            config_path,
+            os.path.join("windows", "gui"),
+            symbol_filename,
```
```suggestion
            context=context,
            config_path=config_path,
            sub_path=os.path.join("windows", "gui"),
            filename=symbol_filename,
```

ikelos commented on volatility3/framework/plugins/windows/poolscanner.py:
```diff
@@ -418,6 +462,40 @@ def generate_pool_scan(
+
+        # repeat the symbol table to match the original `generate_pool_scan` behaviour
+        yield from cls.generate_pool_scan_extended(
+            context, layer_name, symbol_table, symbol_table, constraints
```
I thought part of this PR was to fix names like `layer_name` and `symbol_table` or are you keeping them to avoid changing the API?

ikelos commented on volatility3/framework/plugins/windows/poolscanner.py:
```diff
@@ -331,11 +362,12 @@ def builtin_constraints(
-        layer_name: str,
-        symbol_table: str,
+        kernel_layer_name: str,
+        kernel_symbol_table: str,
```
This is a `kernel_symbol_table_name`, not a `kernel_symbol_table`

ikelos commented on volatility3/framework/plugins/windows/poolscanner.py:
```diff
@@ -331,11 +362,12 @@ def builtin_constraints(
-        symbol_table: str,
+        kernel_layer_name: str,
+        kernel_symbol_table: str,
+        object_symbol_table: str,
```
Same for this

ikelos commented on volatility3/framework/plugins/windows/windowstations.py:
```diff
@@ -0,0 +1,236 @@
+        session_map = cls.get_session_map(context, kernel_module_name, gui_table_name)
+
+        for result in poolscanner.PoolScanner.generate_pool_scan_extended(
+            context,
+            kernel.layer_name,
+            kernel.symbol_table_name,
+            gui_table_name,
+            constraints,
```
```suggestion
            context=context,
            kernel_layer_name=kernel.layer_name,
            kernel_symbol_table=kernel.symbol_table_name,
            object_symbol_table=gui_table_name,
            constraints=constraints,
```
Ideally, this would be updated to `kernel_symbol_table_name` and `object_symbol_table_name` although actually, for the extended one, we probably want `kernel_module_name` instead of the symbol and layer names taken separately...

atcuno replied:
See above, it will after this PR. There will be a PR just to cleanly overhaul the poolscanner APIs and callers.

ikelos replied:
I'd still at least apply the suggested changes to use keyword args (and also, please tag the specific ticket that this'll get fixed by, so we can track it).

ikelos commented on volatility3/framework/symbols/windows/extensions/__init__.py:
```diff
@@ -874,6 +874,12 @@ def _walk_ldr_list(
             sym_table = self.get_symbol_table_name()
+            # Fixes #1636
+            try:
+                peb.Ldr
```
Too many things in this PR.  Next time, please include little one offs like this in their own PR.

atcuno replied:
My bad. We left it here as it was causing many backtraces across multiple plugins since the dlllists get enumerated by many.

ikelos replied:
Still worth a separate PR, just means it can get fixed without being held up by anything else, and much easier to process when not buried in hundreds of other changes.  5:)

dgmcdona replied:
This has been split into its own commit in case we want to break it out.

ikelos replied:
Ok, commit are still harder to split out than just its own PR, but it's a start, thanks.  5:)

ikelos commented on volatility3/framework/symbols/windows/extensions/gui.py:
```diff
@@ -0,0 +1,127 @@
+from volatility3.framework.symbols.windows.extensions import pool
+
+
+class tagWINDOWSTATION(objects.StructType, pool.ExecutiveObject):
```
Should all of these classes be defined in a Versionable class, so that the APIs can be updated?  I realize it's a little new, but we should probably make the change at some point.  Happy for it to be part of a new PR, but if that is how we're going to do it, best to do it now at the start (plugins applying it then need to do a version check on it).  I haven't thought through the consequences of doing this, so it will need discussion before it happens.

atcuno replied:
I rather do it as a new PR (for parity) to understand this system better. Can you link me to an example where this is done?

ikelos replied:
I'm not sure we've strictly ever done what I'm suggesting yet, but it would be like the [`tainting`](https://github.com/volatilityfoundation/volatility3/blob/develop/volatility3/framework/symbols/linux/utilities/tainting.py) code only the main class would contain subclasses.  The idea is that you can version check the whole module.  I haven't figured out whether it's better to version class the module or a class within the module (the class allows us to give it an interface, the module makes it slightly less unusual/easier to read)?  I'm up for suggestions, but the idea is to group all the elements together and tag them all with a version number that can be incremented when the classes are changed in a non-backwards compatible way...

atcuno replied:
Resolving here as now a part of: https://github.com/volatilityfoundation/volatility3/issues/1647

ikelos commented on volatility3/framework/symbols/windows/extensions/pool.py:
```diff
@@ -342,6 +348,12 @@ def get_object_header(self) -> "OBJECT_HEADER":
             native_layer_name=self.vol.native_layer_name,
         )
 
+    def get_name(self, symbol_table_name: Optional[str] = None) -> Optional[str]:
```
Is there a way to version this?

atcuno replied:
I can add it if so. Just link me to an example of how to version an extension.

ikelos replied:
Hmmm, I haven't figured one out yet, but it'll be tricky for external plugins/tools to use the `get_name` method without having to tell people "this only works with this version of the framework" (and that would require a framework MINOR bump).  Options include:
* don't make the changes (not ideal because it forces everything to stay the same)
* bump the framework MINOR version and then remember to set that in any plugins that make use of this call
* Find a way to wrap the extension in a `VersionableInterface` which the plugin can then depend on when it needs to use such code.  Since extensions are usually plugged into the symbol table automatically, it *should* be possible to just you a surround class as a means of versioning, but I don't have a concrete example for you to try, you'd be the first...

atcuno replied:
Resolving here as now a part of: https://github.com/volatilityfoundation/volatility3/issues/1647

ikelos commented on volatility3/plugins/windows/registry/certificates.py:
```diff
@@ -69,13 +69,10 @@ def dump_certificate(
-            layer_name=kernel.layer_name,
-            symbol_table=kernel.symbol_table_name,
+            self.config_path,
+            self.config["kernel"],
```
```suggestion
            base_config_path=self.config_path,
            kernel_module_name=self.config["kernel"],
```

ikelos reviewed (changes requested):
First off, whaaay too big a PR.  This had additional plugins, it had sweeping changes across the entire base (all of which ended up needing tweaking), there was a small bugfix snuck in.  I'm gonna start rejecting these if they're not cut up into their appropriate components.  As it is, I can't guarantee I've caught everything, and you guys now have 84 comments to go through...

Technically all of the touched plugins should have a PATCH version bump, but I'm happy to overlook that.

There's a lot of places where keyword arguments were swapped out in favour of positional arguments?  I don't think there's any benefit of that, it just provides less information for people trying to read the source code as to what's going on.  It would make changing names easier in the future, but I'd actually prefer we think through and choose the right ones first time, rather than removing them so that we could mess with them in the future...

This also goes to a lot of effort to replace the more generic newer `VersionRequirement`, with the older overly specific alias of `PluginRequirement`.  We should be going in the other direction please?  Happy to discuss it if you feel otherwise, but `PluginRequirement` is literallly a facade over `VersionRequirement` (which can be used for more than plugins), so you'll need a good argument.  (See issue #1644).

Because this was such a sweeping change, there's at least two plugins that will still need a further pass (`timers` and `KPCRs`) because there's some daft stuff that got in the tree somehow where some methods still take a separate layer name and symbol table without clear reason for why (and I think there's one that takes a kernel_module *and* a layer_name *and* a kernel_symbol_table*).

So some definite changes here.  On the whole I approve of the changes being made and the effort to smarten up the codebase.  Big changes like this are always going to be hard to make across so many existing plugins (and noting that will affect other plugins that make use of them), but that's why this PR should probably have been  several PRs:

* *Only* Changing parameters
* Adding new plugins
* Bugfix

Please do feel free to submit example PRs (in draft if you want to differentiate them) that can break tests, but at least show what you're intending to do once, before applying it hundreds of times across the codebase, only to need all those times tweaking.  I'd far sooner we checked (and allow tests to fail during development) than create a monstrosity like this where bugs will slip through simply because there's too much to check in one go...

atcuno commented:
Thanks for the feedback. Even though its 80 comments most are easy fixes like the function parameter explicit names in calls.

Check slack links after reading this. I left comments on everything that did not require code changes as I have to leave for airport.

Overall, I agree with these:
1. Changing function calls back to keyword arguments.
2. kpcrs and timers have crap APIs. I already left a FIXME on timers in here as fixing these two will be their own PR after this one merged.
3. The poolscanner API to take the kernel module name for kernel scanners (all but GUI) will also be its own PR after this one. This should be an easy PR as its just changing the calls and bumping versions.
4. Changing to VersionRequirement is fine. This must be new as you never told me this in feedback before.

My two pushback points.
1. "there's at least two plugins that will still need a further pass (timers and KPCRs)" <- I think you later realized the issue in a later PR comment as this bad API is already in develop, so we will fix it right after this.

2. Your saying we would leave unfound bugs here - we ran the codebase with this PR active against 400+ Windows samples and running every single plugin against each sample. There are no backtraces caused by our changes from these runs and obviously if we really broke some version dependency then the plugins would not have run. This is the most complete testing for Volatility 3 that there is.

dgmcdona committed:
Windows SSDT: Simplify `build_module_collection` signature

This simplifies the `build_module_collection` signature to take a single
`kernel_module_name` parameter instead of `layer_name` and
`symbol_table_name` parameters, both of which would only ever belong to
the kernel anyway. This will prevent future confusion for consumers of
this method.

Co-authored-by: Andrew Case <andrew@dfir.org>

ikelos commented:
Yep, all looks good, but could you please go back and everywhere you commented with "it'll be fixed in a future PR" could you file an issue explaining exactly what the PR will cover, and add it to the comment?  That'll help make sure that things which need doing don't get lost, you'll be able to write out how you want to solve it and what the solution should be and when the time comes it can be auto-closed by a PR tagged against it.  Just useful housekeeping.

Yeah, I think I was grumpy because the PR was big and killed several hours (started between 12 and 2am and then finished it between 9 and 10) just reviewing the same kinds of changes with a few random extras thrown in that meant I had to pay attention!  5;P  There was a bit where you added a whole new function but with a signature that was similar to the old one, and it seemed odd to push new code that will definitely need revisiting and fixing up when it could've been added here in a fixed state, but otherwise I just want to make sure that our future desires get documented so they don't get lost or forgotten.

I think what I meant was, split up the PRs into smaller bits so that a huge change doesn't get replicated and then need fixing everywhere?  A draft PR can be used to demonstrate the intended fix without needing to pass all the (now extensive and greatly appreciated) test suite.  I just want to make sure that we don't now all-or-nothing PRs because we're afraid of it failing tests.  It likely will never get committed if it fails, but it's still a good medium to review intended code changes, and I don't want us to be afraid of that.  It's much easier to have review and have a conversation in a PR than randomly on branch code.   I just... you've just gotta give me smaller/more consistent PRs to work on please!  5:P  Beasts like that are just too much for me in one go...

atcuno commented:
Agreed! We don't want PRs like this again either :|

I just went through and made new tickets for all the new/extra efforts discussed here and then resolved the comment thread with the link.

David is working on updates now to knock out the bulk of these at once (like adding the parameters), which should get us pretty close to merge ready.

dgmcdona committed:
Windows SSDT: Updates plugin consumers

Updates all plugins that use the `ssdt` module with correct parameters
and updated version requirement values.

dgmcdona committed:
Windows Modules/Modscan: Clean up APIs

This cleans up the APIs for some methods in the modscan/modules plugins
that currently take separate symbol_table_name and layer_name
parameters, when it really makes more sense to just pass in the context
and the kernel module name.

It also updates the pslist plugin requirement version, and uses the
updated method signatures.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows PsList: Add method for listing procs from kernel

This adds a new classmethod, `list_processes_from_kernel`, updates the
`list_processes` method signature to use only the kernel module name and
the context instead of splitting information about the kernel between
the layer_name and symbol_table_name paramters, and does a major version
number increase on the plugin.

Also updates the documentation to reflect pslist method signature
change.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows PsList: Update dependents

This updates all dependents on windows.pslist.PsList that could be
updated without breaking interface changes of their own to use the
latest windows.pslist.PsList plugin version with a simplified method
signature

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows HiveList: Update list_hives method signature

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Hivelist: Update dependents

This updates all plugins that depend on windows.hivelist.HiveList to use
the updated method signature, and bumps their dependency version
accordingly.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Amcache: Update dependency and change method signature

This updates the Windows Amcache plugin to use the latest changes in
Windows HiveList. It requires a major version bump of its own due to a
breaking method signature change, and is therefore in its own commit.

Co-authored-by: Andrew Case <andrew@dfir.org>

atcuno committed:
Windows GUI Plugins: Adds three plugins

This adds the windowstations, desktops, and deskscan plugins, and
removes the gui.py plugin file that was stubbed out in the introductory
work for the effort.

It also adds a new method to the poolscanner class, but does not bump
the poolscanner version number since there is already a major version
number bump going into this PR.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows PEDump: Update PsList dep and change method signature

This updates PEDump to use the latest version of PsList, updates the
dependency version number, changes one of it's own method signatures to
facilitate the pslist change, and does a major bump of its own version
number.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Consoles: Update hivelist dep and change method signature

This updates the windows.consoles.Consoles plugin to use the updated
hivelist method signature, changing one of its own method signatures as
required and doing a major version bump of its own.

Plugins that depend on consoles also have their method calls changed,
and their dependency versions bumped.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows PEDump: Update dependents

This updates all plugins that both depend on PEDump and won't require
breaking changes of their own. They now call the updated method
signature, and have their pedump requirement version bumped
appropriately.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Threads: Change list_process_threads signature

This updates the windows.threads plugin with a method signature change:
`module_name` is now `kernel_module_name` for clarity.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Threads: Update dependents

This updates the debugregisters and suspended threads plugins' threads
requirement with the latest major version bump.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows PESymbols: Updates requirements, changes interface

This updates the PESymbols plugin to use the latest changes from pslist
and modules. This required breaking interface changes of its own
(altered method signatures) and so receives its own major version bump.
Dependents of PESymbols that _don't_ require any breaking changes have
their method calls and version requirement numbers updated here as well.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows PoolScan: Adds method

This updates the poolscanner plugin with an additional method,
`generate_pool_scan_extended`, and does the corresponding minor version
bump.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Modules: Update dependents

This updates the ssdt, truecrypt, drivermodule, driverirp, and verinfo
plugins to use the new method signatures from the Modules plugin, and
updates the requirement version numbers as well.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Indirect Syscalls: Remove unneeded pslist dep

PsList is not actually used in this plugin, and has therefore been
removed.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Shimcachemem: Updates req's, new major version

This updates the ShimcacheMem plugins with the version bumps on the
modules and pslist requirements, updates the corresponding method calls,
and does a major version bump of its own due to a required breaking
method signature change.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows ScheduledTasks: update hivelist dep, breaking change

This updates the ScheduledTasks plugin with an updated hivelist
requirement version number, changes the corresponding method call, and
does a major version bump of its own due to a changed parameter name.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Extensions: Fix traceback when accessing peb.Ldr

This fixes a simple InvalidAddressException traceback by just catching
the exception on the member access and continuing.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Netstat: Update modules req, adds breaking change

This updates the netstat plugin to use the new method signature in the
modules plugin; however, this requires a breaking change of its own
which is done here as well. The major version is bumped accordingly;
this plugin has no dependents that require updates at this time.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows SvcScan: Update reqs, add breaking change

This updates the windows SvcScan plugin with the latest changes from
hivelist/pslist, fixing up method calls and bumping requirement version
numbers. In order to use the new method signatures, breaking changes
were required to the svscan public methods, so a major version bump has
been added.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows svclist: Update svcscan dependency

This updates the svclist plugin with breaking changes to its public
methods in order to update calls to the svcscan methods. Both
requirement and plugin version numbers have been updated accordingly
here.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Orphan Kernel Threads: Update modules dep, add breaking change

This updates the orphan kernel threads plugin to use the latest changes
from the modules plugin, updating method signatures and bumping the
requirement version number. This required breaking interface changes in
the plugin itself, so the major version number has been bumped.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Strings: Update pslist req and add breaking change

This uses the latest changes from windows.pslist, updated the
requirement version number. It also required breaking changes of its
own, so the Strings version number has been given a major version bump
as well.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows SvcDiff: Update reqs, adds breaking change

This updates the svcdiff plugin to use the latest changes from svclist
and svcscan, updating the requirement version numbers, and giving itself
a major version bump due to a changed method signature.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows UnhookedSystemCalls: Update reqs, add breaking change

This update the unhooked system calls plugin to use the latest changes
from pslist, updating the requirement version numbers and bumping its
own major version number due to a changed method signature.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows SuspiciousThreads: Adds missing requirement, updates other

This updates the threads version number to the latest (2.0.0) and adds a
missing plugin requirement for thrdscan.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Windows Suspended Threads: Updates pe_symbols req

This bumps the requirement version number for pe_symbols, and uses the
latest method signature.

Co-authored-by: Andrew Case <andrew@dfir.org>

dgmcdona committed:
Code Review: Fix pslist param

Change to self.current_kernel_name so people can change this value.

dgmcdona committed:
Code Review (style): Use keyword args for clarity

This updates a whole host of method calls to pass keyword arguments
instead of positional arguments.

dgmcdona committed:
Code Review: PluginRequirement -> VersionRequirement

Code Review: PluginRequirement -> Version Requirement

dgmcdona committed:
Windows DeskScan: Add version to class

code review fix

dgmcdona committed:
Code Review: Simplify dictionary construction

Also adds some keywords in method call args, and fixes an incorrect
type-hint.

dgmcdona committed:
Code Review: Parameter renaming

Makes it explicit that these are symbol table names, not actual symbol
tables.

dgmcdona committed:
Code Review: Remove redundant kernel module reconstruction

dgmcdona committed:
CodeQL Fix: Unused variable

Also underscores an unused unpacked tuple value

ikelos commented on volatility3/framework/plugins/windows/envars.py:
```diff
@@ -58,13 +58,11 @@ def _get_silent_vars(self) -> List[str]:
 
         for hive in hivelist.HiveList.list_hives(
-            context=self.context,
+            self.context,
```
```suggestion
            context=self.context,

```
No point leaving this one out...  5;)

ikelos commented on volatility3/framework/plugins/windows/getservicesids.py:
```diff
@@ -69,18 +69,16 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
         # Get the system hive
         for hive in hivelist.HiveList.list_hives(
-            context=self.context,
+            self.context,
```
```suggestion
            context=self.context,

```

ikelos commented on volatility3/framework/plugins/windows/getsids.py:
```diff
@@ -101,14 +101,12 @@ def lookup_user_sids(self) -> Dict[str, str]:
         sids = {}
         for hive in hivelist.HiveList.list_hives(
-            context=self.context,
+            self.context,
```
```suggestion
            context=self.context,

```

ikelos commented on volatility3/framework/plugins/windows/handles.py:
```diff
@@ -393,9 +393,8 @@ def run(self):
-                context=self.context,
-                layer_name=kernel.layer_name,
-                symbol_table=kernel.symbol_table_name,
+                self.context,
```
```suggestion
                context=self.context,
```

ikelos commented on volatility3/framework/plugins/windows/modules.py:
```diff
@@ -92,12 +92,13 @@ def _generator(self):
             session_layers = list(
                 self.get_session_layers(
-                    self.context, kernel.layer_name, kernel.symbol_table_name
+                    self.context,
```
```suggestion
                    context=self.context,
```

ikelos commented on volatility3/framework/plugins/windows/modules.py:
```diff
@@ -268,26 +308,29 @@ def find_session_layer(
-        kvo = context.layers[layer_name].config.get("kernel_virtual_offset", None)
+        kernel = context.modules[kernel_module_name]
+
+        kvo = context.layers[kernel.layer_name].config.get(
```
At some point this should become `kernel.offset`, but I don't think we can guarantee that's the case yet.  One to remember though...

ikelos commented on volatility3/framework/plugins/windows/pslist.py:
```diff
@@ -206,35 +208,37 @@ def filter_func(x):
+
         # We only use the object factory to demonstrate how to use one
-        kvo = context.layers[layer_name].config.get("kernel_virtual_offset", None)
+        kvo = context.layers[kernel.layer_name].config.get(
```
I think we can just check `kernel.offset` rather than looking up the `kvo`?  It's fine to leave it this way, but we'll probably rip it out as a mechanism down the road.

ikelos commented on volatility3/framework/plugins/windows/registry/getcellroutine.py:
```diff
@@ -28,27 +28,24 @@ def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]
         # is inside of the kernel (ntoskrnl)
         for hive_object in hivelist.HiveList.list_hives(
-            context=self.context,
+            self.context,
```
```suggestion
            context=self.context,
```

ikelos commented on volatility3/framework/plugins/windows/registry/userassist.py:
```diff
@@ -295,18 +295,16 @@ def _generator(self):
         # get all the user hive offsets or use the one specified
         for hive in hivelist.HiveList.list_hives(
-            context=self.context,
+            self.context,
```
```suggestion
            context=self.context,
```

ikelos commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -1135,17 +1135,16 @@ def get_software_hive(
         return next(
             hivelist.HiveList.list_hives(
-                context=context,
+                context,
```
```suggestion
                context=context,
```

ikelos commented on volatility3/framework/plugins/windows/skeleton_key_check.py:
```diff
@@ -672,9 +670,8 @@ def run(self):
-                    context=self.context,
-                    layer_name=kernel.layer_name,
-                    symbol_table=kernel.symbol_table_name,
+                    self.context,
```
```suggestion
                    context=self.context,
```

ikelos commented on volatility3/framework/plugins/windows/svcdiff.py:
```diff
@@ -61,10 +62,12 @@ def service_diff(
+
         if not symbols.symbol_table_is_64bit(
-            context, symbol_table
+            context, kernel.symbol_table_name
```
```suggestion
            context=context, symbol_table=kernel.symbol_table_name
```
No reason for these two to be different

ikelos commented on volatility3/framework/plugins/windows/svclist.py:
```diff
@@ -60,16 +62,17 @@ def _get_exe_range(cls, proc) -> Optional[Tuple[int, int]]:
+
         if not symbols.symbol_table_is_64bit(
-            context, symbol_table
+            context, kernel.symbol_table_name
```
```suggestion
            context=context, symbol_table=kernel.symbol_table_name
```

ikelos commented on volatility3/framework/plugins/windows/svcscan.py:
```diff
@@ -144,15 +144,15 @@ def _create_service_table(
+
         for hive in hivelist.HiveList.list_hives(
-            context=context,
+            context,
```
```suggestion
            context=context,
```

dgmcdona committed:
Windows GUI Symbols: JSON Lint

ikelos commented on volatility3/framework/symbols/windows/extensions/pool.py:
```diff
@@ -326,12 +326,18 @@ class ExecutiveObject(interfaces.objects.ObjectInterface):
 
-    def get_object_header(self) -> "OBJECT_HEADER":
+    def get_object_header(
+        self, symbol_table_name: Optional[str] = None
```
Technically an additive change, but I think there's enough gone on in this PR that we could probably do with a MINOR version bump of the framework anyway please.  At least until #1647 gets resolved.

ikelos reviewed (commented):
Ok, looking much better, thanks for clearing up all those bits!  Just minor bits to get it over the line...  5:D

dgmcdona commented:
I'll get the rest of that cleaned up. I also reworked the commit history so things are as atomic as I could make them - might help catch things that would otherwise go unnoticed regarding all of the requirement/plugin version updates. I think there were one or two that I picked up as I was going through. The bugfixes for scheduled tasks and the `peb.Ldr` traceback are in their own commits as well, as are all of the GUI additions.

dgmcdona committed:
Framework: Minor version bump

There are a lot of things going into this so we're doing a minor
framework version bump.

dgmcdona committed:
Code Review: Get rid of 'kvo' in favor of kernel.offset

dgmcdona committed:
Code Review: Remove unneeded kernel module reconstruction

dgmcdona commented:
@ikelos I addressed all of the comments and fixed up all of the arguments where they should be passed as keyword arguments, so I think this is good to go

dgmcdona commented:
> I'll get the rest of that cleaned up. I also reworked the commit history so things are as atomic as I could make them - might help catch things that would otherwise go unnoticed regarding all of the requirement/plugin version updates. I think there were one or two that I picked up as I was going through. The bugfixes for scheduled tasks and the `peb.Ldr` traceback are in their own commits as well, as are all of the GUI additions.

bumping this just in case it got lost in all of the force-pushed commits

ikelos commented on volatility3/framework/layers/registry.py:
```diff
@@ -66,7 +66,7 @@ def __init__(
         # if it exists and update RegistryHive._base_layer
         for proc in pslist.PsList.list_processes(
-            self.context, self.config["base_layer"], self.config["nt_symbols"]
+            context=self.context, kernel_module_name=self.config["kernel_module_name"]
```
Just want to get a little clarity on this, where does `kernel_module_name` get defined as the parameter this takes?  I would've expected a `requirements` method or something, but it doesn't seem to have one and that makes me worry that it's just convention that we look for that particular config name.  It might be right, but it will need fixing (at the very least an issue so we don't forget to investigate it).

It and the dependency on pslist needs to go too.

atcuno replied:
I added it to here as the code path through hivelist is what creates a `subconfig` that the registry layer uses:

https://github.com/volatilityfoundation/volatility3/blob/a553ee39212e189e6866044ab80ecb5190f1a1aa/volatility3/framework/plugins/windows/registry/hivelist.py#L181

This is also the code path that will need to be updated to satisfy the ticket we made about the registry layer walking the process list (it all boils down to the same root cause in hivelist.py):

https://github.com/volatilityfoundation/volatility3/issues/1645

So all the related hivelist code will be cleaned up in 1645 anyway.

ikelos replied:
Ok, thanks, that was the bit that was missing for me.  5:)

ikelos commented on volatility3/framework/plugins/windows/windowstations.py:
```diff
@@ -0,0 +1,237 @@
+            sub_path=os.path.join("windows", "gui"),
+            filename=symbol_filename,
+            class_types=gui.class_types,
+            native_types=native_types,
```
Shouldn't we be taking the `symbol_table`'s existing native types from windows, rather than the static ones we built years ago?  I believe the inbuilt nativetypes are an absolute fallback, not expected to be used in production except in dire emergency.

dgmcdona replied:
We ran into an issue around `f32` when using the natives from windows in some older Win7 samples. The natives types for modern windows are the following:
```python
{'HRESULT',
 'array',
 'bitfield',
 'bytes',
 'char',
 'double',
 'enum',
 'f32',
 'function',
 'int',
 'long',
 'long long',
 'pointer',
 'short',
 'string',
 'unsigned char',
 'unsigned int',
 'unsigned long',
 'unsigned long long',
 'unsigned short',
 'void',
 'wchar'}
 ```

But we came across a Windows 7 Service Pack 0 x64 sample that has the following native types:
```python
{'array',
 'bitfield',
 'bytes',
 'char',
 'double',
 'enum',
 'function',
 'int',
 'long',
 'long long',
 'pointer',
 'short',
 'string',
 'unsigned char',
 'unsigned int',
 'unsigned long',
 'unsigned long long',
 'unsigned short',
 'void',
 'wchar'}
```

So we couldn't create an intermediate symbol table for the GUI types without the `f32` in the native types. [This statement](https://github.com/volatilityfoundation/volatility3/blob/5f0c2b5eeb6dcd3340571924ed582ebf51c502a6/volatility3/framework/symbols/intermed.py#L463) was evaluating to `False`, leading to an exception being raise [here](https://github.com/volatilityfoundation/volatility3/blob/5f0c2b5eeb6dcd3340571924ed582ebf51c502a6/volatility3/framework/symbols/intermed.py#L496). Using the statically defined native types worked in this instance since it guarantees we'll always have `f32`.

Is there a way to update the Windows native types to include f32 if it's missing? I didn't see an obvious way to do it.

ikelos replied:
Awww, no, not gracefully there isn't, and adding one would likely be more disruptive that just using the nativetables.  Could you please add some comments in, because at some point we probably ought to deprecate those native tables, and we'll need to know why we've bothered using it there...

ikelos replied:
Actually, does it *need* to be in the native table?  You might just be able to define it as a normal type instead, and those can be added/modified?  I'm just thinking because I don't think we call our native type f32, I think we call it float, so I figure something weird must go on anyway?

dgmcdona replied:
It turns out just removing the `native_types` argument fixes this, since the required natives will be constructed from the base types here (at least I think so) if you don't provide them: https://github.com/volatilityfoundation/volatility3/blob/caa52c08d34e1a2af8d438062ce69164db730594/volatility3/framework/symbols/intermed.py#L690

From what I've seen so far, removing the parameter fixes the issue regardless of whether or not the windows types have the f32 native type defined.

ikelos reviewed (commented):
Much better, and nothing show stopping, but a lot of bugs that need following up on please.  I'll wait to hear about about the native types stuff before I push it.

ikelos reviewed (commented):
Just the two more points to answer/check but fine ok they're resolved.

dgmcdona committed:
Code Review: Don't use to-be-deprecated static natives

This was originally done to solve a problem where `f32` wasn't available
in the native types from every kernel version. However, it ended up not
being necessary - we can just omit `native_types` from the method call,
and it will construct the types as-needed using the definition for
`float` in `base_types` from the JSON files.

ikelos commented:
So will this now fail on certain old versions of windows?

dgmcdona committed:
Windows Versions: Fix broken OSDistinguisher

Something happened when picking _EPROCESS members before that caused
this to not function properly. This one relies on a more stable type
removal instead of _EPROCESS members.

ikelos commented:
Bah, I messed up the merge, that needs removing instead, sorry!  5:S

ikelos commented on volatility3/framework/plugins/linux/check_idt.py:
```diff
@@ -88,6 +88,10 @@
             self.context, vmlinux.name, modules
         )
 
+        is_32bit = not symbols.symbol_table_is_64bit(
```
Wow, that's kinda creepy...  5:S

dgmcdona commented:
> So will this now fail on certain old versions of windows?

It shouldn't, since the native types are constructed by `_get_natives` now.

dgmcdona commented:
@ikelos Hope you don't mind, I rolled back the bad merge, rebased onto the latest develop, and force-pushed

ikelos commented:
Yep, all good.  Sorry I was trying to fix it myself through the web gui, but I've got a relatively high miss rate trying that, sorry!  5:P
